File: tools/enrich.mjs
// Enrich data/artifacts.json from the British Museum object pages.
//
// This mirrors the proven recipe in your harvest.ts (real Chrome, headed,
// anti-automation args, a warmed persistent profile, a fixed render wait),
// generalised to every object in our dataset. It fills ONLY blank fields
// (material, date_text, description) and trims descriptions to 2 sentences.
//
// Reuse your already-Cloudflare-warmed profile and existing Playwright install:
//
//   cd ~/bm-test
//   BM_PROFILE=/Users/davidwaite/.gemini/antigravity/scratch/british-museum-extractor/chrome_data_extractor \
//   NODE_PATH=/Users/davidwaite/.gemini/antigravity/scratch/british-museum-extractor/node_modules \
//   node tools/enrich.mjs --dump Y_EA77434     # 1. sanity check one object
//   ...same env... node tools/enrich.mjs --limit 20 --dry   # 2. preview 20
//   ...same env... node tools/enrich.mjs                     # 3. full run -> writes artifacts.json
//
// Flags: --limit N | --dry | --dump BM_ID | --headless
// It's resumable: blanks-only + checkpoints every 25, so stop/restart freely.
import { chromium } from "playwright";
import fs from "fs";
import path from "path";
import { fileURLToPath } from "url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ARTIFACTS = process.env.BM_ARTIFACTS || path.join(__dirname, "..", "data", "artifacts.json");
const PROFILE = process.env.BM_PROFILE || path.join(__dirname, ".enrich_profile_chrome");
const OBJECT_URL = (id) => `https://www.britishmuseum.org/collection/object/${id}`;
const DESC_MAX = 300;

const argv = process.argv.slice(2);
const flag = (name) => argv.includes(name);
const opt = (name, d) => { const i = argv.indexOf(name); return i >= 0 ? argv[i + 1] : d; };
const LIMIT = parseInt(opt("--limit", "0"), 10);
const DRY = flag("--dry");
const DUMP = opt("--dump", null);
const HEADLESS = flag("--headless");

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

function twoSentences(text) {
  text = (text || "").replace(/\s+/g, " ").trim();
  if (!text) return "";
  const parts = text.split(/(?<=[.!?])\s+/);
  let out = parts.slice(0, 2).join(" ").trim();
  if (out.length > DESC_MAX) out = out.slice(0, DESC_MAX).replace(/\s+\S*$/, "").replace(/[,;:]$/, "") + "…";
  return out;
}

// Parse the object page's <dt>/<dd> detail list (runs in the browser).
async function parseDetails(page) {
  return await page.evaluate(() => {
    const data = {};
    for (const dt of Array.from(document.querySelectorAll("dt"))) {
      const key = (dt.textContent || "").trim();
      if (!key) continue;
      const parts = [];
      let el = dt.nextElementSibling;
      while (el && el.tagName.toLowerCase() === "dd") {
        const t = (el.textContent || "").replace(/\s+/g, " ").trim();
        if (t) parts.push(t);
        el = el.nextElementSibling;
      }
      if (parts.length && !(key in data)) data[key] = parts.join(", ");
    }
    return data;
  });
}

function mapFields(d) {
  return {
    description: twoSentences(d["Description"] || ""),
    date_text: (d["Production date"] || d["Date"] || d["Cultures/periods"] || "").trim(),
    material: (d["Materials"] || d["Material"] || "").trim(),
  };
}

async function warm(page) {
  await page.goto("https://www.britishmuseum.org/collection/search?keyword=museum&has_image=1",
    { waitUntil: "domcontentloaded", timeout: 60000 });
  await sleep(5000);
  try {
    const btn = page.locator('button:has-text("Allow all cookies"), button:has-text("Reject all cookies")').first();
    if (await btn.isVisible({ timeout: 2500 })) { await btn.click(); await sleep(1500); }
  } catch { /* no modal */ }
}

async function fetchObject(page, id) {
  await page.goto(OBJECT_URL(id), { waitUntil: "domcontentloaded", timeout: 45000 });
  await sleep(6000); // let the detail list + carousel render (matches harvest.ts)
  let html = await page.content();
  if (!html.includes("object-detail__data-term")) {
    await sleep(4000);
    html = await page.content();
  }
  if (!html.includes("object-detail__data-term")) return null; // CF/empty — retry next run
  return mapFields(await parseDetails(page));
}

function needsEnrich(a) {
  return !(a.material && a.date_text && a.description);
}

async function main() {
  const data = JSON.parse(fs.readFileSync(ARTIFACTS, "utf8"));
  const ctx = await chromium.launchPersistentContext(PROFILE, {
    channel: "chrome",
    headless: HEADLESS,
    viewport: { width: 1280, height: 900 },
    ignoreDefaultArgs: ["--enable-automation"],
    args: ["--disable-blink-features=AutomationControlled", "--test-type",
           "--hide-crash-restore-bubble", "--no-sandbox"],
  });
  const page = ctx.pages()[0] || (await ctx.newPage());

  try {
    await warm(page);

    if (DUMP) {
      const fields = await fetchObject(page, DUMP);
      if (!fields) { console.log(`No detail fields for ${DUMP} (Cloudflare/empty).`); return; }
      console.log("=== mapped (what would be written) ===");
      console.log(JSON.stringify(fields, null, 2));
      return;
    }

    let todo = data.filter((a) => needsEnrich(a) && a.bm_id);
    if (LIMIT) todo = todo.slice(0, LIMIT);
    console.log(`${todo.length} objects still have blanks; processing${DRY ? " (dry run)" : ""}…`);

    const filled = { material: 0, date_text: 0, description: 0 };
    let noRecord = 0;
    for (let i = 0; i < todo.length; i++) {
      const a = todo[i];
      let fields = null;
      try {
        fields = await fetchObject(page, a.bm_id);
      } catch (e) {
        console.log(`  [${i + 1}/${todo.length}] ${a.bm_id}: ${e.message}`);
      }
      if (!fields) { noRecord++; }
      else {
        for (const f of ["material", "date_text", "description"]) {
          if (!a[f] && fields[f]) { a[f] = fields[f]; filled[f]++; }
        }
      }
      if ((i + 1) % 25 === 0) {
        console.log(`  …${i + 1}/${todo.length}  filled: ${JSON.stringify(filled)}  misses: ${noRecord}`);
        if (!DRY) fs.writeFileSync(ARTIFACTS, JSON.stringify(data, null, 1));
      }
      await sleep(1200); // polite throttle
    }

    console.log(`\nDone. Filled: ${JSON.stringify(filled)}. No record for ${noRecord}.`);
    if (DRY) console.log("Dry run — artifacts.json NOT written.");
    else { fs.writeFileSync(ARTIFACTS, JSON.stringify(data, null, 1)); console.log(`Wrote ${ARTIFACTS}`); }
  } finally {
    await ctx.close();
  }
}

main().catch((e) => { console.error("[fatal]", e); process.exit(1); });

File: tools/enrich.py
#!/usr/bin/env python3
"""
Enrich data/artifacts.json from the British Museum collection website.

Network access is required, so run this on a machine that can reach
britishmuseum.org (e.g. your Mac) — NOT inside the Claude sandbox.

Behaviour (per your settings):
  - ONLY fills blank fields (material, date_text, description). Never
    overwrites a value that is already present.
  - description is trimmed to at most 2 sentences (and a hard char cap).
  - Resumable + polite: every fetched page is cached under
    tools/.enrich_cache/, so re-runs skip work and you can stop/restart.

Typical use:
  1. Validate the field mapping on one object first:
       python3 tools/enrich.py --dump Y_EA77434
     (prints the raw record the parser found — check it has the right
      description/date/material, then paste it to me if anything's off.)
  2. Dry-run a handful:
       python3 tools/enrich.py --limit 20 --dry-run
  3. Full run (writes data/artifacts.json):
       python3 tools/enrich.py

Flags:
  --limit N     only process the first N objects that still have blanks
  --dry-run     fetch + parse + report, but don't write artifacts.json
  --dump ID     fetch one bm_id, print the parsed record, exit
  --delay S     seconds between live fetches (default 1.0; be kind)
"""
import argparse
import gzip
import json
import re
import sys
import time
import urllib.request
import urllib.error
from pathlib import Path

BASE = Path(__file__).resolve().parent.parent
DATA = BASE / "data" / "artifacts.json"
CACHE = BASE / "tools" / ".enrich_cache"
OBJECT_URL = "https://www.britishmuseum.org/collection/object/{bm_id}"
UA = ("Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 "
      "(KHTML, like Gecko) Version/17.0 Safari/605.1.15")
DESC_MAX_CHARS = 300

# Fields we are allowed to fill, mapped to the candidate keys we look for in
# the British Museum page record. Edit the candidate lists if --dump shows the
# real keys differ.
FIELD_CANDIDATES = {
    "description": ["description", "Description", "physicalDescription", "comment"],
    "date_text": ["date", "productionDate", "production_date", "Date", "dateText"],
    "material": ["material", "materials", "Materials", "Material", "medium"],
}


# --------------------------------------------------------------- networking

def fetch(url):
    req = urllib.request.Request(url, headers={
        "User-Agent": UA,
        "Accept": "text/html,application/xhtml+xml",
        "Accept-Encoding": "gzip, deflate",
        "Accept-Language": "en-GB,en;q=0.9",
    })
    with urllib.request.urlopen(req, timeout=30) as r:
        raw = r.read()
        if r.headers.get("Content-Encoding") == "gzip":
            raw = gzip.decompress(raw)
    return raw.decode("utf-8", "replace")


def get_page(bm_id, delay):
    """Return the page HTML, from cache if present, else fetch + cache."""
    CACHE.mkdir(parents=True, exist_ok=True)
    cached = CACHE / f"{bm_id}.html"
    if cached.exists():
        return cached.read_text("utf-8", "replace"), True
    url = OBJECT_URL.format(bm_id=bm_id)
    last = None
    for attempt in range(4):
        try:
            html = fetch(url)
            cached.write_text(html, "utf-8")
            time.sleep(delay)
            return html, False
        except urllib.error.HTTPError as e:
            if e.code == 404:
                cached.write_text("", "utf-8")  # remember the miss
                return "", False
            last = e
        except Exception as e:  # noqa: BLE001 - network is messy; retry
            last = e
        time.sleep(2 ** attempt)
    raise RuntimeError(f"failed to fetch {url}: {last}")


# --------------------------------------------------------------- parsing

def extract_next_data(html):
    """Pull the embedded __NEXT_DATA__ JSON blob the BM site ships."""
    m = re.search(
        r'<script id="__NEXT_DATA__"[^>]*>(.*?)</script>', html, re.DOTALL)
    if not m:
        return None
    try:
        return json.loads(m.group(1))
    except json.JSONDecodeError:
        return None


def walk(node):
    """Yield every dict in a nested JSON structure."""
    if isinstance(node, dict):
        yield node
        for v in node.values():
            yield from walk(v)
    elif isinstance(node, list):
        for v in node:
            yield from walk(v)


def find_record(blob, bm_id, museum_number):
    """Heuristically locate the object's detail dict within the page blob."""
    best, best_score = None, -1
    needles = {bm_id, (museum_number or "").replace(" ", "")}
    for d in walk(blob):
        text = json.dumps(d, ensure_ascii=False)
        score = 0
        if any(n and n in text for n in needles):
            score += 2
        # Looks like an object record if it carries several detail-ish keys.
        score += sum(1 for keys in FIELD_CANDIDATES.values()
                     for k in keys if k in d)
        if score > best_score:
            best, best_score = d, score
    return best if best_score >= 2 else None


def as_text(value):
    """Flatten a string / list / nested label dict into plain text."""
    if value is None:
        return ""
    if isinstance(value, str):
        return value.strip()
    if isinstance(value, list):
        return ", ".join(t for t in (as_text(v) for v in value) if t)
    if isinstance(value, dict):
        for k in ("label", "name", "title", "value", "text"):
            if k in value:
                return as_text(value[k])
    return ""


def pick(record, field):
    for key in FIELD_CANDIDATES[field]:
        if key in record:
            txt = as_text(record[key])
            if txt:
                return txt
    return ""


def two_sentences(text):
    text = re.sub(r"\s+", " ", text).strip()
    if not text:
        return ""
    parts = re.split(r"(?<=[.!?])\s+", text)
    out = " ".join(parts[:2]).strip()
    if len(out) > DESC_MAX_CHARS:
        out = out[:DESC_MAX_CHARS].rsplit(" ", 1)[0].rstrip(",;:") + "…"
    return out


def parse_object(html, bm_id, museum_number):
    blob = extract_next_data(html)
    if blob is None:
        return None
    record = find_record(blob, bm_id, museum_number)
    if record is None:
        return None
    return {
        "description": two_sentences(pick(record, "description")),
        "date_text": pick(record, "date_text"),
        "material": pick(record, "material"),
        "_record": record,
    }


# --------------------------------------------------------------- main

def needs_enrich(a):
    return not (a.get("material") and a.get("date_text") and a.get("description"))


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--limit", type=int, default=0)
    ap.add_argument("--dry-run", action="store_true")
    ap.add_argument("--dump", metavar="BM_ID")
    ap.add_argument("--delay", type=float, default=1.0)
    args = ap.parse_args()

    data = json.loads(DATA.read_text("utf-8"))

    if args.dump:
        html, cached = get_page(args.dump, args.delay)
        if not html:
            print(f"No page for {args.dump} (404 or empty).")
            return
        a = next((x for x in data if x.get("bm_id") == args.dump), {})
        parsed = parse_object(html, args.dump, a.get("museum_number"))
        if not parsed:
            print("Could not locate the object record in __NEXT_DATA__.")
            print("Paste this script's output to Claude so the mapping can be fixed.")
            return
        rec = parsed.pop("_record")
        print("=== extracted fields ===")
        print(json.dumps(parsed, indent=2, ensure_ascii=False))
        print("\n=== raw record keys ===")
        print(sorted(rec.keys()))
        print("\n=== raw record (first 2000 chars) ===")
        print(json.dumps(rec, indent=2, ensure_ascii=False)[:2000])
        return

    todo = [a for a in data if needs_enrich(a) and a.get("bm_id")]
    if args.limit:
        todo = todo[:args.limit]
    print(f"{len(todo)} objects still have blanks; processing "
          f"{len(todo)}{' (dry run)' if args.dry_run else ''}…")

    filled = {"material": 0, "date_text": 0, "description": 0}
    no_record = 0
    for i, a in enumerate(todo, 1):
        try:
            html, cached = get_page(a["bm_id"], args.delay)
        except Exception as e:  # noqa: BLE001
            print(f"  [{i}/{len(todo)}] {a['bm_id']}: fetch error: {e}")
            continue
        parsed = parse_object(html, a["bm_id"], a.get("museum_number")) if html else None
        if not parsed:
            no_record += 1
        else:
            for field in ("material", "date_text", "description"):
                if not a.get(field) and parsed.get(field):
                    a[field] = parsed[field]
                    filled[field] += 1
        if i % 50 == 0:
            print(f"  …{i}/{len(todo)}  filled so far: {filled}")

    print(f"\nDone. Filled: {filled}. No record found for {no_record} objects.")
    if args.dry_run:
        print("Dry run — data/artifacts.json NOT written.")
    else:
        DATA.write_text(json.dumps(data, ensure_ascii=False, indent=1), "utf-8")
        print(f"Wrote {DATA}")


if __name__ == "__main__":
    main()
